gradient/src/main.rs

```diff
@@ -0,0 +1,164 @@
+    let mut first_way = true;
+    println!("Reading {osm_path}");
+    let nodes_progress = ProgressBar::new_spinner().with_style(
+        ProgressStyle::with_template("[{elapsed_precise}] {human_len} nodes read ({per_sec})")
+            .unwrap(),
+    );
+    let ways_progress = ProgressBar::new_spinner().with_style(
+        ProgressStyle::with_template("[{elapsed_precise}] {human_len} ways read ({per_sec})")
+            .unwrap(),
+    );
+    osm_reader::parse(&fs_err::read(osm_path)?, |elem| match elem {
+        Element::Node { id, lon, lat, .. } => {
+            nodes_progress.inc(1);
+            node_mapping.insert(id, Coord { x: lon, y: lat });
+        }
+        Element::Way { id, node_ids, tags } => {
+            if tags.contains_key("highway") && tags.get("area") != Some(&"yes".to_string()) {
+                if first_way {
+                    nodes_progress.finish();
+                    first_way = false;
+                }
+                ways_progress.inc(1);
+                highways.push((id, node_ids));
+            }
+        }
+        Element::Relation { .. } | Element::Bounds { .. } => {}
+    })?;
+    ways_progress.finish();
+
+    Ok(split_edges(node_mapping, highways))
+}
+
+fn split_edges(
+    node_mapping: HashMap<NodeID, Coord>,
+    ways: Vec<(WayID, Vec<NodeID>)>,
+) -> Vec<(WayID, LineString)> {
+    println!("Splitting ways into edges");
+
+    // Count how many ways reference each node
+    let mut node_counter: HashMap<NodeID, usize> = HashMap::new();
+    for (_, node_ids) in &ways {
+        for node in node_ids {
+            *node_counter.entry(*node).or_insert(0) += 1;
+        }
+    }
+
+    // Split each way into edges
+    let progress = ProgressBar::new(ways.len() as u64).with_style(ProgressStyle::with_template(
+        "[{elapsed_precise}] [{wide_bar:.cyan/blue}] {human_pos}/{human_len} ({per_sec}, {eta})").unwrap());
+    let mut edges = Vec::new();
+    for (way_id, node_ids) in ways {
+        progress.inc(1);
+        let mut pts = Vec::new();
+
+        let num_nodes = node_ids.len();
+        for (idx, node) in node_ids.into_iter().enumerate() {
+            pts.push(node_mapping[&node]);
+            // Edges start/end at intersections between two ways. The endpoints of the way also
+            // count as intersections.
+            let is_endpoint =
+                idx == 0 || idx == num_nodes - 1 || *node_counter.get(&node).unwrap() > 1;
+            if is_endpoint && pts.len() > 1 {
+                edges.push((way_id, LineString::new(std::mem::take(&mut pts))));
+
+                // Start the next edge
+                pts.push(node_mapping[&node]);
+            }
+        }
+    }
+    progress.finish();
+    edges
+}
+
+fn get_steepness(edge: &LineString, elevation: &mut GeoTiffElevation<BufReader<File>>) -> isize {
+    // Calculate height at the endpoints. Ignore anything in the middle -- if a long edge has a
```

Comment:
Can get feedback about how this layer will be used, and how important this is. Most edges are pretty short, but maybe there are some long rural ones out there to be more careful about